public/app/features/scenes/components/app/SceneApp.tsx

```diff
@@ -0,0 +1,167 @@
+import React from 'react';
+import { Route, Switch, useRouteMatch } from 'react-router-dom';
+
+import { NavModelItem } from '@grafana/data';
+import { SceneComponentProps, SceneObject, SceneObjectBase, SceneObjectStatePlain } from '@grafana/scenes';
+import { Page } from 'app/core/components/Page/Page';
+
+import { getLinkUrlWithAppUrlState, useAppQueryParams } from '../../apps/utils';
+
+export interface SceneAppState extends SceneObjectStatePlain {
+  pages: SceneAppPage[];
+}
+
+export interface SceneRouteMatch<Params extends { [K in keyof Params]?: string } = {}> {
+  params: Params;
+  isExact: boolean;
+  path: string;
+  url: string;
+}
+
+export interface SceneAppRoute {
+  path: string;
+  page?: SceneAppPage;
+  drilldown?: SceneAppDrilldownView;
+}
+
+export class SceneApp extends SceneObjectBase<SceneAppState> {
+  public static Component = ({ model }: SceneComponentProps<SceneApp>) => {
+    const { pages } = model.useState();
+
+    return (
+      <Switch>
+        {pages.map((page) => (
+          <Route key={page.state.url} exact={false} path={page.state.url}>
+            {page && <page.Component model={page} />}
+          </Route>
+        ))}
+      </Switch>
+    );
+  };
+}
+
+export interface SceneAppPageState extends SceneObjectStatePlain {
+  title: string;
+  url: string;
+  routePath?: string;
+  subTitle?: string;
+  hideFromBreadcrumbs?: boolean;
+  tabs?: SceneAppPage[];
+  getScene: (routeMatch: SceneRouteMatch) => SceneObject;
+  drilldowns?: SceneAppDrilldownView[];
+  getParentPage?: () => SceneAppPage;
+  preserveUrlKeys?: string[];
+}
+
+const sceneCache = new Map<string, SceneObject>();
+
+export class SceneAppPage extends SceneObjectBase<SceneAppPageState> {
+  public static Component = ({ model }: SceneComponentProps<SceneAppPage>) => {
+    const { tabs } = model.state;
+
+    if (tabs) {
+      const routes: React.ReactNode[] = [];
+
+      for (const page of tabs) {
+        routes.push(
+          <Route key={page.state.url} exact={true} path={page.state.routePath ?? page.state.url}>
+            <page.Component model={page} />
+          </Route>
+        );
+
+        if (page.state.drilldowns) {
+          for (const drilldown of page.state.drilldowns) {
+            console.log('registering drilldown route', drilldown.routePath);
+            routes.push(
+              <Route key={drilldown.routePath} exact={false} path={drilldown.routePath}>
```

Comment:
Why the drill-down route is non-exact?

Reply:
@dprokop to support nested routes, not exactly sure why but without it the second level routing does not work.

For example we have drilldown
```
  drilldowns: [
          {
            routePath: '/scenes/grafana-monitoring/handlers/:handler',
```

and the defines a SceneAppPage that has tabs like

```
new SceneAppPage({
        title: 'Logs',
        url: baseUrl + '/logs',
        routePath: '/scenes/grafana-monitoring/handlers/:handler/logs',
        getScene: () => getHandlerLogsScene(handler),
        preserveUrlKeys: ['from', 'to', 'var-instance'],
        drilldowns: [
          {
            routePath: '/scenes/grafana-monitoring/handlers/:handler/logs/:secondLevel',
            getPage: getSecondLevelDrilldown,
          },
        ],
```